storage/buffer.go

```diff
@@ -607,7 +607,20 @@ func addSample(s chunks.Sample, buf []chunks.Sample, r *sampleRing) []chunks.Sam
 		}
 	}
 
-	buf[r.i] = s
+	switch spec := s.(type) {
+	case hSample:
+		if spec.h != nil {
+			spec.h = spec.h.Copy()
+		}
+		buf[r.i] = spec
+	case fhSample:
+		if spec.fh != nil {
+			spec.fh = spec.fh.Copy()
+		}
+		buf[r.i] = spec
+	default:
+		buf[r.i] = s
+	}
```

Comment:
How about adding a `Copy()` method to the `chunks.Sample` interface and then just write
```Go
buf[r.i] = s.Copy()
```
here, at the price, of course, of implementing that method for all four implementations.
The type switch looks awkward.

(I hacked together this approach. It's actually quite simple. If you like, I can push it here.)